### ISSUE TITLE
Fail on large 95GB Restore Backup (Restore Virtual Servers)

### ISSUE THREAD

**Issue by chris001** [virtualmin/virtualmin-gpl#17]
Steps to reproduce:
1. Make full backup on cPanel server according to steps on `virtualmin.com`.
2. Login as root to new Virtualmin, current version, running on 16.04 OS, 1GB RAM, no swap, 500 GB disk.
3. Click Backup and Restore, Restore Backup, Local file, Select file in file navigator ( `/root/backup-10.24.2017_16-47-37_cPanelUsername.tar.gz` ).
4. Click `Show What Will Be Restored`.
5. See the spinning circle in the top right corner of the virtualmin page, as it is running the process.
6. Notice CPU usage is showing activity on `htop` for processes `gzip`, `cat`, `tar`, and `restore.cgi`.
7. After some time, probably 15 minutes, CPU activity stops, and browser shows some error, `Connection reset`.

**Comment by chris001** [comment]
Fail also with `virtualmin migrate-domain`
`virtualmin migrate-domain --type cpanel --source /root/backup-10.24.2017_16-47-37_cPanelUsername.tar.gz --domain mydomain.com --user cPanelUsername --pass mypassword --webmin --test`

```
Validating migration file ..
.. found domain mydomain.com user cPanelUsername password m
```
And it hangs there until the `ssh` session disconnects due to timeout.

Is it possible the perl script for `migrate-domain` is malfunctioning because the cPanel password for the full backup file contains a special character `$` ??

**Comment by chris001** [comment]
```
Error 
Failed to migrate virtual server : Expecting form-data encoding, but got normal encoding
Return to previous page
```

**Comment by jcameron** [comment]
The browser-based restore might be failing due to your browser's connection to the server timing our due to inactivity. However, the command-line restore should work..

Does the output from `restore-domain` really halt just after "password m" ?

**Comment by chris001** [comment]
Yes the output really halted after the first letter.  The second letter of the password was the `$` character.  There was a total of about 10 characters in the password.  Should I put quotes around the password when it contains special symbols such as `$` just to be sure that the perl code doesn't treat it as a variable injection or something?

**Comment by chris001** [comment]
Here's the results from another run, slightly different parameters:
```
virtualmin migrate-domain --type cpanel --source /root/backup-10.24.2017_16-47-37_cPanelUsername.tar.gz --user cPanelUsername --pass "cPanelPassword" --webmin
Validating migration file ..
.. done

Starting migration of yourDomain.com from local file /root/backup-10.24.2017_16-47-37_cPanelUsername.tar.gz ..

Checking for cPanel features ..
.. found Home directory, Administration user, Log file rotation, Webmin login, BIND DNS domain, Mail for domain.

However, the follow features are not supported or enabled on your system : Apache website, Webalizer reporting, MySQL database. Some functions of the migrated virtual server may not work.

Checking for clashes and dependencies ..
Log file rotation cannot be enabled unless a website is

Migration failed! See the error output above.
```

Looks like the virtualmin server is running Nginx and no database.  The Apache and MySQL features on Virtualmin are disabled, so Virtualmin throws an error, it can't migrate the Apache and MySQL stuff from the cPanel full backup.

Would be great if `virtualmin migrate-server` would ask you if you want to enable the disabled features (Apache and MySQL) so that the `migrate-server` command could complete successfully!

**Comment by chris001** [comment]
After manually enabling the features, the following command for `migrate-domain` worked:
`virtualmin migrate-domain --type cpanel --source /root/backup-10.24.2017_16-47-37_cPanelUsername.tar.gz --user cPanelUsername --pass "cPanelPassword" --webmin`
However, the `migrate-domain.pl` script threw AT LEAST TWO `perl` errors to fix. Maybe more `perl` errors, don't know because `screen` doesn't let you scroll back beyond the current single terminal page of text:
```
Use of uninitialized value in numeric eq (==) at /usr/share/webmin/bind8/bind8-lib.pl line 546.
    .. done

    Adding to email domains list ..
    .. done

    Adding new virtual website ..
    .. done

    Adding webserver user www-data to server's group ..
    .. done

    Performing other Apache configuration ..
    .. done

    Setting up scheduled Webalizer reporting ..
    .. done

    Setting up log file rotation ..
    .. done

    Creating MySQL login ..
    .. done

    Creating Webmin user ..
Use of uninitialized value in split at /usr/share/webmin/acl/acl-lib.pl line 625.
    .. done

    Re-starting DNS server ..
    .. done

    Applying web server configuration ..
    .. not running!
```

**Comment by jcameron** [comment]
Those perl warnings are pretty harmless (but we will fix them).

The deeper issue of the cPanel backup not restoring unless Apache and MySQL are enabled is harder to fix. The cPanel backup actually contains Apache directives from the original system, which can't be reliably migrated to Nginx.

**Comment by chris001** [comment]
Yes I'm OK with Apache and MySQL features having to be enabled in Virtualmin, so that a cPanel backup containing Apache and MySQL features/directives can be restored successfully.

The only issue is the amount of huge amount of time and so many retries it took to get the `migrate-domain` to work.  
* Each try took at least an hour for the script to finally throw errors and die.  
* The first 3 tries where `Restore Backup` from the web interface, crashed, it left behind GBs of temp files in `/tmp/.webmin/` - ideally it'd be best if the script when it crashes, it would clean up all those temp files.
* Getting all the way to the end of the run, at least an hour, then complaining that the features are not enabled - it'd be best if `migrate-domain` would show these warnings when the features required are not enabled, up front at the start of the run.
* The password (and probably username also) must be `"quoted"` in double-quotes when they contain special characters, anything other than numbers or letters, to avoid crashing the perl script.

**Comment by jcameron** [comment]
The long time needed before reporting an error is unfortunately hard to avoid, because Virtualmin doesn't know what features a cPanel backup will use until it has been downloaded and extracted - and those are the slow parts of the operation.

Quoting is out of our control, as it's the shell that interprets special characters before Virtualmin's script is even run.

However, not cleaning up temp files seems like a bug. Does this happen even when a migration from the command line fails?

**Comment by chris001** [comment]
* `migrate-server.pl` from the command line was good about cleaning up after its crashes.  When I mistakenly used "import virtual server" from the web interface, specified the cpanel full backup file, and clicked "show what will be imported" (instead of "migrate virtual server") that is what crashed and left behind GB of files in `/tmp/.webmin`.
* About the quoting of passwords on the command line. It would be good to at least mention double quotes are recommended for passwords containing anything other than alphanumeric, in the command line `--help`. Even better, halt the perl script immediately when you detect the user passes in a password without `"double quotes"` around it.
* All of these crashes were with the `.tar.gz` file coming directly from the local filesystem, and they all took ages to come back with errors and crash.  It'd be great if it could do a quick scan of the archive, when it's a local file, detect which features needed, then prompt the user for an OK to enable any currently disabled features needed to restore the backup successfuly - apache, mysql, etc.  If not that, then at least warn the user, when they want to migrate a cPanel backup, they better have Apache MySQL and Webalizer enabled, or whatever features their cPanel account was using, because if those features are not enabled, the migrate will fail, and they'll lose all of that time, possibly hours.

**Comment by jcameron** [comment]
I'll see if I can re-produce this issue of leftover temp files - normally when any Virtualmin script exits, it cleans up all leftover files. However, a hard crash (ie. out of RAM) may make this impossible.

Suggesting that passwords be quoted in the help output is a good idea - I'll do that.

Unfortunately, a quick scan of a tar.gz file is impossible. But for features like Apache that we know 100% will be required for migrating a cPanel domain, I can add a check as soon as migration starts.

**Comment by chris001** [comment]
Maybe a dumb question, but, here goes.  
* Let's say, you go into Virtualmin and "Add Server / Migrate Virtual Server/ CPanel full backup", and then do the exact same thing "Add Server / Migrate Virtual Server / CPanel full backup"...
* Is the "Migrate cpanel" process additive??  Or, does the latest "Migrate cpanel" overwrite/clobber the previous "Migrate cpanel?  
* Trying  to ask before actually doing it, in the hopes of saving many hours of wasted upload/download time of 100GB of cpanel backups.

**Comment by jcameron** [comment]
If the first migrate got far enough to create the domain, the second one will fail with an error reporting that it already exists.

**Comment by chris001** [comment]
OK so is `domain already exists` the only case where a repeated Migrate CPanel backup will fail?
For example, let's say Migrate zip 1 contains domains A, B, C and Migrate zip 2 contains domains D, E, and F, since there are no already existing domains, the Migrate zip 2 will not fail with error, it will simply add its domains to the Virtualmin server ?

**Comment by jcameron** [comment]
Correct - migrations will not fail if there is no domain name clash.

**Comment by chris001** [comment]
Would it be wise for webmin to detect the cpanel backup contains apache2/mysql/postgresql/any other features, determine whether or not webmin has all the features enabled, and if they're disabled, to automatically re-enable the the features, so that the `migrate-domain` process will continue to work and not fail.

**Comment by jcameron** [comment]
Were these features disabled globally on your system, on the Features and Plugins page?

**Comment by jcameron** [comment]
Enabling features globally as part of the restore process seems risky .. but it sounds like what you really want is to be notified of this problem because the entire 95 GB backup is extracted? 

Unfortunately, based on the format of the cPanel backups, I can't see any way to actually implement this :-(
